Data/HashMap/Internal/Array.hs

```diff
@@ -474,6 +478,27 @@ fromList n xs0 =
     go (x:xs) mary i = do write mary i x
                           go xs mary (i+1)
 
+fromList' :: Int -> [a] -> Array a
+fromList' n xs0 =
+    CHECK_EQ("fromList'", n, Prelude.length xs0)
+        run $ do
+            mary <- new_ n
+            go xs0 mary 0
+  where
+    go [] !mary !_   = return mary
+    go (!x:xs) mary i = do write mary i x
+                           go xs mary (i+1)
```

Comment:
What's the motivation for adding this in addition to the existing `fromList` which looks identical at first glance?

Reply:
It's strict in the elements, so the resulting `HashMap` won't be full of thunks.

Reply:
Ah! Can you add haddocks that point this out?

Reply:
Ah, but the leaves may still be thunks .... Hmm.... Maybe it's better to be lazy after all?

Reply:
> Maybe it's better to be lazy after all?

@madgen, what do you think / prefer?

Reply:
@sjakobi in my particular application laziness doesn't play much role, so I prefer it there weren't full of thunks, but I don't think that is a superior choice in general. Do what you think would be most generally applicable please.

Reply:
@sjakobi, The problem is that it's not a "full of thunks" vs. "not full of thunks". If we're lazy in the leaves, as we really should be, then we should be lazy in the spines as well for best results. But users may want to be strict in the leaves, so we should have a strict lift in the `Strict` module.

Reply:
I don't understand. What you mean by leaves. Elements in the `HashMap`? Isn't the assumption that they will be floated out and made static data if they can?

(That's why I'd like `SmallArray#` literals, you'll just know that `lifted` expressions won't have any thunks as it perfectly would be just constructors or literals all the way).

Reply:
@phadej, I guess we want to force construction of the values? Will the derived instance do that? I got a bit confused again.